File: javaweb/loja-virtual/src/br/com/alura/model/Produto.java
package br.com.alura.model;

public class Produto {
	
	private Integer id;
	private String nome;
	private String descricao;
	
	public Produto(Integer id, String nome, String descricao) {
		this.id = id;
		this.nome = nome;
		this.descricao = descricao;
	}
	
	public Produto(String nome, String descricao) {
		this.nome = nome;
		this.descricao = descricao;
	}
	
	public Produto() {
	}
	
	@Override
	public String toString() {
		return this.getId() + ", " + this.getNome() + ", " + this.getDescricao();
	}

	public Integer getId() {
		return id;
	}

	public void setId(Integer id) {
		this.id = id;
	}

	public String getNome() {
		return nome;
	}

	public void setNome(String nome) {
		this.nome = nome;
	}

	public String getDescricao() {
		return descricao;
	}

	public void setDescricao(String descricao) {
		this.descricao = descricao;
	}
}

File: php-oo/entity/Produto.php
<?php

  require_once 'Categoria.php';

  abstract class Produto
  {
      private $id;
      private $nome;
      private $preco;
      private $descricao;
      private $categoria;
      private $usado;
      private $tipoProduto;

      public function calculaDesconto($desconto = 0.1)
      {
          if ($desconto > 0 && $desconto < 1) {
              $this->preco -= $this->preco * $desconto;
          }
          return $this->preco;
      }

      public function getId()
      {
          return $this->id;
      }

      public function setId($id)
      {
          $this->id = $id;
      }

      public function getNome()
      {
          return $this->nome ;
      }

      public function setNome($nome)
      {
          $this->nome = $nome;
      }

      public function getPreco()
      {
          return $this->preco;
      }

      public function setPreco($preco)
      {
          $this->preco = $preco;
      }

      public function getDescricao()
      {
          return $this->descricao;
      }

      public function setDescricao($descricao)
      {
          $this->descricao = $descricao;
      }

      public function getCategoria()
      {
          return $this->categoria;
      }

      public function setCategoria(Categoria $categoria)
      {
          $this->categoria = $categoria;
      }

      public function getUsado()
      {
          return $this->usado ;
      }

      public function setUsado($usado)
      {
          $this->usado = $usado;
      }

      public function getTipoProduto()
      {
          return $this->tipoProduto ;
      }

      public function setTipoProduto($tipoProduto)
      {
          $this->tipoProduto = $tipoProduto;
      }

      public function isLivro()
      {
          return $this instanceof Livro;
      }

      public function temTaxaImpressao() {
        return $this instanceof LivroFisico;
      }

      public function temWaterMark() {
        return $this instanceof Ebook;
      }

      public function calculaImposto()
      {
          return $this->getPreco() * 0.195;
      }

      public abstract function atualizaBaseadoEm($params);
  }
